File: challenge_2/scala/ndsmith3/SingleNumber.scala
object SingleNumber {
  def unique(elems: Array[Any], soln: List[Any] = Nil): List[Any] = {
    elems.length match {
      case 0 => soln
      case _ if elems.tail.contains(elems.head) =>
        unique(elems.filter(_ != elems.head), soln)
      case _ => unique(elems.tail, elems.head :: soln)
    }
  }
}

File: challenge_2/elixir/joegotflow83/single_number/lib/single_number.ex
defmodule SingleNumber do
  def unique(input) do
    input
    |> Enum.reduce(%{}, &count/2)
    |> Map.to_list
    |> find_uni
  end

  def find_uni([head | tail]) do
    case head do
      {num, 1} ->
        num
      _ ->
        find_uni(tail)
    end
  end

  def find_uni([]) do
    "There are no unique elements."
  end

  def count(num, acc) do
    Map.update acc, num, 1, &(&1 + 1)
  end
end
